Source/LinqToDB/Common/OptionsContainer.cs

```diff
@@ -96,20 +96,24 @@ public TSet FindOrDefault<TSet>(TSet defaultOptions)
 			return Find<TSet>() ?? defaultOptions;
 		}
 
+		readonly object _sync = new ();
+
 		/// <summary>
 		/// Returns options set by set type <typeparamref name="TSet"/>. If options doesn't contain specific options set, it is created and added to options.
 		/// </summary>
 		/// <typeparam name="TSet">Options set type.</typeparam>
 		/// <returns>
 		/// Returns options set by set type <typeparamref name="TSet"/>. If options doesn't contain specific options set, it is created and added to options.
 		/// </returns>
+		[Pure]
 		public virtual TSet Get<TSet>()
 			where TSet : class, IOptionSet, new()
 		{
 			if (Find<TSet>() is { } set)
 				return set;
 
-			(_sets ??= new())[typeof(TSet)] = set = new();
+			lock (_sync)
+				_sets = new (_sets ?? []) { [typeof(TSet)] = set = new() };
```

Comment:
I like this better, but with a `lock`, we don't need a new assignment every time.
```suggestion
			lock (_sync)
				(_sets ??= new())[typeof(TSet)] = set = new();
```

Reply:
Yeah the approach with a lock is safe as long as we publish a new `Dictionary` and don't mutate the existing one (which would require a lock on readers as well). 👍

Reply:
If we do not assign it every time, we need to use lock everywhere we read `_sets` which is much worse.
Again, this code will be called once for Option Container and all its copies. Once for each exotic option for the entire application run-time.

Reply:
but do we really need "volatile"?